drivers: usb: dwc2: esp32: Add support

raffarost opened the pull request:
Add USB device driver support for ESP32 devices with USB-OTG peripheral.

Commit:
drivers: usb: dwc2: esp32: Add support

Add USB device driver support for ESP32 devices with
USB-OTG peripheral.

Signed-off-by: Raffael Rostagno <raffael.rostagno@espressif.com>

Commit:
dts: esp32s3: Add USB-OTG support

Add USB-OTG peripheral support to ESP32S3.

Signed-off-by: Raffael Rostagno <raffael.rostagno@espressif.com>

jfischer-no commented on drivers/usb/udc/udc_dwc2.c:
```diff
@@ -3160,21 +3179,7 @@ static const struct udc_api udc_dwc2_api = {
 		k_thread_name_set(&priv->thread_data, dev->name);		\
 	}									\
 										\
-	static void udc_dwc2_irq_enable_func_##n(const struct device *dev)	\
-	{									\
-		IRQ_CONNECT(DT_INST_IRQN(n),					\
-			    DT_INST_IRQ(n, priority),				\
-			    udc_dwc2_isr_handler,				\
-			    DEVICE_DT_INST_GET(n),				\
-			    DW_IRQ_FLAGS(n));					\
-										\
-		irq_enable(DT_INST_IRQN(n));					\
-	}									\
-										\
-	static void udc_dwc2_irq_disable_func_##n(const struct device *dev)	\
-	{									\
-		irq_disable(DT_INST_IRQN(n));					\
-	}									\
+	UDC_DWC2_IRQ_DT_INST_DEFINE(n)						\
```
Looks like a stray change that was not justified in the commit message. Anyway, I do not agree with it, you should use irq_enable()/irq_disable().

jfischer-no replied:
I do not understand the problem with irq_enable/irq_disable. How was it accepted not to use generic API for your platform? Are you working on fixing it?

raffarost replied:
I did some checking, and the current behavior for ESP Xtensa chips is: Zephyr's API is used to control non-muxed interrupt lines (such as wifi, systimer). Most peripherals such as USB allow routing sources through an interrupt matrix and for this, the interrupt controller API is used. I'm not sure how much this deviates from the standard implementation, but I believe significant changes would be needed to otherwise accommodate this.

sylvioalves replied:
@jfischer-no, for esp32 device, we still need the irq calls in there, otherwise it is no-go for now. Would you review it again?

jfischer-no replied:
I still cannot find an explanation as to why your platform cannot use generic functions, or why this has not been done all the time is supported in Zephyr.

raffarost replied:
ESP chips use an interrupt controller peripheral to route interrupt sources coming from each peripheral to the CPU's IRQ lines. For this reason we implement an interrupt controller driver and call it's API directly from all ESP drivers. A few peripherals have their interrupt source lines connected directly to the CPU IRQ's, thus having dedicated lines (e.g. wifi). As USB OTG is routed using the Interrupt Matrix peripheral we need to connect it's interrupt source to the IRQ accessing HAL LL calls. As this can't be accomplished using Zephyr's `irq_connect()` without making significant changes to the way we manage interrupts, we are asking to treat interrupt allocation as a `quirk` for the time being. We are already discussing internally if we should rework the interrupt system implementation and we did a review recently, so this will still be an ongoing work for us. Also, we are still not implementing static interrupt allocation even for IRQ lines that are static and don't need routing.

jfischer-no replied:
> @jfischer-no if you grep the `esp_intr_alloc`, you'll find it is used 30 times - that means it is used in every single one driver related to ESP32xx so far - that makes me think about how it made it to the sources so far?

That is a good question. Also, why was the ESP32xx platform included in the project when it is not using the basic common API?

> We would be more than happy to use the Zephyr IRQ API, but unfortunately it is not suitable for ESP32 SoCs, yet. The task of supporting Zephyr IRQ API is undergoing now, and it is not the most trivial one. Hope you can see the bigger picture now and re-evaluate your `-1` opinion, which is blocking this work for several months for now. Thanks!

It was not the only point that delayed this PR and I already offered to escalate it, see https://github.com/zephyrproject-rtos/zephyr/pull/85600#discussion_r2267086157.

marekmatej replied:
> This is not limited to this hardware and can be seen in many SoCs of the same architecture, but across the tree as well.

This is exactly the point that was overlooked here.

jfischer-no replied:
> The change for request is not directly related to the feature being added and
is unfortunatly how things are done in all drivers of this hardware family.

That is not correct. It is directly related to the feature, as it extends driver's internal vendor quirk API, and it would not be needed if ESP32xx platform would use generic IRQ API. If that gets accepted, then there will be no reason to use generic API by anyone else.

sylvioalves replied:
> @sylvioalves or anyone else working on this platform, can you open a bug and track progress toward cleaning up interrupt handling?
> 
https://github.com/zephyrproject-rtos/zephyr/issues/95817

jfischer-no commented on west.yml:
```diff
@@ -162,7 +162,7 @@ manifest:
         - hal
     - name: hal_espressif
-      revision: 07526d59121822a7d85de6c1a4adcf47b906d232
+      revision: pull/411/head
```
This DWC2 UDC driver is vendor independent. There is no need to pull all the madness in your HAL or use CONFIG_UDC_DRIVER there.

marekmatej replied:
Obviously, The HAL has all the low-level drivers for USB. Or are there any ESP32 USB stuff in the Zephyr sources we aren't aware of?

jfischer-no replied:
Before you commented, did you read my comment. Did you understand the driver design?

marekmatej replied:
I recommend looking at what is pulled in HAL before making bold statements. I see you don't like to use HAL implementations of what is already available in Zephyr and that's a fair objection. However, the HAL is pulling in files needed for PHY and USB clock settings, and that's vendor specific.

raffarost replied:
we are aware that some files in HAL are not being used, and we might need to work on removing them

jfischer-no commented on drivers/usb/udc/udc_dwc2_vendor_quirks.h:
```diff
@@ -311,6 +311,169 @@ DT_INST_FOREACH_STATUS_OKAY(QUIRK_NRF_USBHS_DEFINE)
+		return -ENODEV;
+	}
+
+	int ret = clock_control_on(cfg->clock_dev, cfg->clock_subsys);
```
The variable definition should always be at the top of a block.

jfischer-no commented on drivers/usb/udc/udc_dwc2.c:
```diff
@@ -44,10 +44,10 @@ enum dwc2_drv_event_type {
+ * of OUT endpoints.
  */
-#define UDC_DWC2_GRXFSIZ_FS_DEFAULT	(15U + 512U/4U)
+#define UDC_DWC2_GRXFSIZ_FS_DEFAULT (15U + 2 * (64U/4U + 1))
```
Largest endpoint MPS is not 64.

raffarost replied:
can you help figuring out how to properly size the FIFO?
our device has FIFO depth of 200 words, and I couldn't figure out how to properly downsize it

tmon-nordic replied:
Is your PHY supporting only Full-Speed?

tmon-nordic replied:
The current FIFO assignment code is completely broken in the DWC2 driver IMHO. I think it cannot be really fixed without providing the driver with information about available device configurations (how many endpoints, what type and with what maximum packet size). Such information would be necessary upfront because once the device is enabled, the RX fifo cannot be resized (and similarly once IN endpoint is enabled, the TX fifo cannot be resized).

Only if such full information was somehow exposed, then the driver could determine how much space it can assign to each endpoint. If there would be not enough FIFO to facilitate all endpoints, the driver could then decide to e.g. enable thresholding only for IN endpoints, or for all endpoints.

This is not that much hitting issue for current Nordic Semiconductor devices because the FIFO is large enough for most projects. However, with smaller SPRAM configurations, this is much more pressing issue (like here for esp32).

If you have an idea how to modify the stack to enable the driver to know this information, I would gladly review it.

sylvioalves replied:
@tmon-nordic, your suggestion is super valid but I think it could be handled in a separate PR for now. Would you review it again and let us know whether you think re-work is needed? We do want this to move forward.

roma-jam replied:
How about open the driver just with the EP0 only configuration for RX FIFO (means max 64 for FS, and 512 for HS), and then to change this limit during the others EP enabling? 
 
Anyway, the periodic USB packet could be 1024 bytes for FS, so this case is also not covered in the current macros...

If this is a default macros which calculates the depth based on the speed, maybe we can make it a bit clear, something like: 

```
#define UDC_DWC2_GRXFSIZ_DEFAULT(mps) (15U + 2 * ((mps)/4U + 1))
```

and then: 

```
if (hs_phy) {
	default_depth = UDC_DWC2_GRXFSIZ_DEFAULT(512);
} else {
	default_depth = UDC_DWC2_GRXFSIZ_DEFAULT(64);
}
```

tmon-nordic replied:
> How about open the driver just with the EP0 only configuration for RX FIFO (means max 64 for FS, and 512 for HS), and then to change this limit during the others EP enabling?

The problem with dynamic FIFO reallocation is that all EPEna bits must be 0 and all NAKSts bits must be 1. DOEPCTL0.EPEna cannot be disabled after enabled by application.

> Anyway, the periodic USB packet could be 1024 bytes for FS, so this case is also not covered in the current macros...

Isochronous Full-Speed transfer is up to 1023 bytes.

jfischer-no commented on samples/subsys/usb/cdc_acm/boards/esp32s3_devkitm_procpu.overlay:
```diff
@@ -0,0 +1,14 @@
+/*
```
Obviously none of these board overlays are acceptable. But I am a bit surprised, did it not confuse you why there are no board overlays like in the samples?

marekmatej replied:
There are plenty of .overlay files for STM `nucleo` and Nordic boards hidden under the `samples/subsys/usb/...` Those overlays should be correctly placed under the `boards` folder too. 
However, what better location would you suggest for putting the endpoint nodes related to each USB example?

raffarost commented on drivers/usb/udc/udc_dwc2_vendor_quirks.h:
```diff
@@ -311,6 +311,169 @@ DT_INST_FOREACH_STATUS_OKAY(QUIRK_NRF_USBHS_DEFINE)
+                                                                                                   \
+	static int esp32_usb_otg_init_##n(const struct device *dev)                                \
+	{                                                                                          \
+		return esp32_usb_otg_init(dev, &usb_otg_config_##n, &usb_otg_data_##n);            \
+	}                                                                                          \
+                                                                                                   \
+	static int esp32_usb_otg_enable_phy_##n(const struct device *dev)                          \
+	{                                                                                          \
+		return esp32_usb_otg_enable_phy(&phy_ctx_##n, true);                               \
+	}                                                                                          \
+                                                                                                   \
+	static int esp32_usb_otg_disable_phy_##n(const struct device *dev)                         \
+	{                                                                                          \
+		return esp32_usb_otg_enable_phy(&phy_ctx_##n, false);                              \
+	}                                                                                          \
+                                                                                                   \
+	struct dwc2_vendor_quirks dwc2_vendor_quirks_##n = {                                       \
+		.init = esp32_usb_otg_init_##n,                                                    \
+		.post_enable = esp32_usb_otg_enable_phy_##n,                                       \
+		.disable = esp32_usb_otg_disable_phy_##n,                                          \
+		.irq_clear = NULL,                                                                 \
+	};
+
+#define UDC_DWC2_IRQ_DT_INST_DEFINE(n)                                                             \
+	static void udc_dwc2_irq_enable_func_##n(const struct device *dev)                         \
+	{                                                                                          \
+		esp_intr_enable(usb_otg_data_##n.int_handle);                                      \
+	}                                                                                          \
+                                                                                                   \
+	static void udc_dwc2_irq_disable_func_##n(const struct device *dev)                        \
+	{                                                                                          \
+		esp_intr_disable(usb_otg_data_##n.int_handle);                                     \
+	}
+
+DT_INST_FOREACH_STATUS_OKAY(QUIRK_ESP32_USB_OTG_DEFINE)
+
+#undef DT_DRV_COMPAT
+
+#endif /*DT_HAS_COMPAT_STATUS_OKAY(espressif_esp32_usb_otg) */
+
```
is it the formatting? (clang was used here)
can you point the issues?

raffarost commented:
@jfischer-no
would you mind clarifying these points so I can work on the PR? thank you

Commit:
boards: esp32s3: Add label for USB-OTG (UDC0)

Add default zephyr_udc0 label for USB-OTG.

Signed-off-by: Raffael Rostagno <raffael.rostagno@espressif.com>

tmon-nordic commented on include/zephyr/drivers/usb/udc.h:
```diff
@@ -46,6 +46,8 @@ struct udc_device_caps {
 	uint32_t can_detect_vbus : 1;
 	/** Maximum packet size for control endpoint */
 	enum udc_mps0 mps0 : 2;
+	/** Maximum packet size for data endpoints */
+	uint16_t mps : 11;
```
What about other drivers? UDC header is common for all drivers.

raffarost replied:
Good point. If this approach is feasible, should we change the other drivers?

tmon-nordic replied:
I think there must be some other way to implement what you want without touching the global UDC header. All endpoints already do have the max packet size stored in `udc_ep_config`.

Is the limit to 64 due to actual hardware constraints? What is your configuration for Largest Rx Data FIFO Depth and Largest Device Mode Tx Data FIFO x Depth (x in your case seems to be 0 to 6)? Or is it entirely to avoid "allocating too much" to individual endpoints and then running out of SPRAM?

tmon-nordic replied:
The failure is actually is resolved by changing the `default_depth`, not by limiting the `mps` in `udc_ep_config`.

I think the proposal here is too invasive. Maybe it could be for the time being solved with `#define MAX_RXFIFO_GDFIFO_PERCENTAGE 25` which would limit how much of total available SPRAM (`GDFIFOCfg` field in `GDFIFOCFG` register) can be used by RxFIFO. The only change would then be just capping `priv->rxfifo_depth` to minimum of whatever value is currently calculated and what the max percentage of available SPRAM is. This could then be changed later to e.g. enable thresholding when DMA is enabled if the information about endpoints is known upfront (see https://github.com/zephyrproject-rtos/zephyr/pull/85600#discussion_r1995653967)

tmon-nordic commented on samples/subsys/usb/hid-keyboard/app.overlay:
```diff
@@ -13,3 +13,7 @@
 };
+
+&zephyr_udc0 {
+	status = "okay";
```
Why in sample and not in board like every other target?

raffarost replied:
As we are sharing the PHY with USB/JTAG functionality, I'm afraid we can't leave USB enabled by default in the board file. JTAG wouldn't work for builds that don't use USB, and an external transceiver would be necessary then.

Commit:
drivers: usb: dwc2: Limit RX FIFO size allocation

Limit RX FIFO size allocation to a percentage of the total
SPRAM available, to cap allocation according to memory size.

Signed-off-by: Raffael Rostagno <raffael.rostagno@espressif.com>

tmon-nordic commented on drivers/usb/udc/udc_dwc2.c:
```diff
@@ -65,6 +65,9 @@ enum dwc2_drv_event_type {
 /* Get Data FIFO access register */
 #define UDC_DWC2_EP_FIFO(base, idx)	((mem_addr_t)base + 0x1000 * (idx + 1))
 
+/* Percentage limit of how much SPRAM can be allocated for RXFIFO */
```
Other comments do use `RxFIFO` capitalization which matches the official name used in Programming Guide.

raveious commented:
There are boards outside of Espressif vendor folder that use the ESP32-S3. You should probably update those to also disable OTG, just to keep their existing behavior consistent.

raffarost commented:
> There are boards outside of Espressif vendor folder that use the ESP32-S3. You should probably update those to also disable OTG, just to keep their existing behavior consistent.

USB OTG is disabled by default for all boards, including Espressif's. This PR only adds `zephyr_udc0` label to the reference boards (while keeping disabled) to facilitate running samples, and these can optionally be added to the other boards as well.

roma-jam commented on drivers/usb/udc/udc_dwc2.c:
```diff
@@ -65,6 +65,9 @@ enum dwc2_drv_event_type {
 #define UDC_DWC2_EP_FIFO(base, idx)	((mem_addr_t)base + 0x1000 * (idx + 1))
 
+/* Percentage limit of how much SPRAM can be allocated for RxFIFO */
+#define MAX_RXFIFO_GDFIFO_PERCENTAGE 25
```
When we introduced a maximum limit in percentage, maybe it could be better to make this constant configurable (I tried it with several values and seems that we can set the range `25-75`). WDYT ? 
 
Something like: 

```suggestion
#define MAX_RXFIFO_GDFIFO_PERCENTAGE CONFIG_UDC_DRIVER_RXFIFO_THRESHOLD
```

with the changes in `zephyr/drivers/usb/udc/Kconfig` :
```
config UDC_DRIVER_RXFIFO_USAGE_LIMIT
	int "RX FIFO usage limit (%)"
	range 25 75
	default 25
	help 
	  Limits how much of total available SPRAM (GDFIFOCFG field in GDFIFOCFG
	  register) can be used by RxFIFO.
```

tmon-nordic replied:
Why 25 to 75? If this ends up as Kconfig, then 100% should be the default (disable the limit altogether, which should remove the code that puts the limit in place) because the DWC2 Kconfig options are shared by all targets. You could then add different default based on your target.

roma-jam replied:
> then 100% should be the default 

Yes, agreed. 

May I ask is there any available Nordic board (dev kit) that can easily be used to verify usb applications?

roma-jam reviewed (approved):
Ok, I tried to check the FIFO changes, but seems that I need more time to make something more valuable.

phy part LGTM.


One small thing: there is no protection for the `esp32_usb_otg_enable_phy` on multicore systems, but maybe this is not a problem as I don't see it in the other quirks. Sorry if I am missing something.

raffarost commented:
> Ok, I tried to check the FIFO changes, but seems that I need more time to make something more valuable.
> 
> phy part LGTM.
> 
> One small thing: there is no protection for the `esp32_usb_otg_enable_phy` on multicore systems, but maybe this is not a problem as I don't see it in the other quirks. Sorry if I am missing something.

thank you, @roma-jam :)

jfischer-no commented on west.yml:
```diff
@@ -162,7 +162,7 @@ manifest:
         - hal
     - name: hal_espressif
-      revision: dbc28ad4c1bdcdb25e79ca225cb5528a75d8dc91
+      revision: 772c1c69d282c0fbbdf3a5002c4f3e0dcc2111df
```
This still pulls unnecessary changes/code and is not addressed.

jfischer-no commented on drivers/usb/udc/udc_dwc2_vendor_quirks.h:
```diff
@@ -315,6 +315,173 @@ DT_INST_FOREACH_STATUS_OKAY(QUIRK_NRF_USBHS_DEFINE)
+	}
+
+	/* pinout config to work in USB_OTG_MODE_DEVICE */
+	esp_rom_gpio_connect_in_signal(GPIO_MATRIX_CONST_ONE_INPUT, USB_OTG_IDDIG_IN_IDX, false);
+	esp_rom_gpio_connect_in_signal(GPIO_MATRIX_CONST_ONE_INPUT, USB_SRP_BVALID_IN_IDX, false);
+	esp_rom_gpio_connect_in_signal(GPIO_MATRIX_CONST_ONE_INPUT, USB_OTG_VBUSVALID_IN_IDX,
+				       false);
+	esp_rom_gpio_connect_in_signal(GPIO_MATRIX_CONST_ZERO_INPUT, USB_OTG_AVALID_IN_IDX, false);
+
+	if (cfg->phy_ctx->target == USB_PHY_TARGET_INT) {
+		gpio_set_drive_capability(USBPHY_DM_NUM, GPIO_DRIVE_CAP_3);
+		gpio_set_drive_capability(USBPHY_DP_NUM, GPIO_DRIVE_CAP_3);
+	}
```
Can it use pinctrl if not why?

raffarost replied:
As we are adding support initially for OTG using only internal transceiver, pins are fixed (cannot be configured freely).
Later, when adding config for external transceiver support, we might use `pinctrl` indeed.

jfischer-no commented on drivers/usb/udc/udc_dwc2_vendor_quirks.h:
```diff
@@ -315,6 +315,173 @@ DT_INST_FOREACH_STATUS_OKAY(QUIRK_NRF_USBHS_DEFINE)
+					      (phy_ctx->target == USB_PHY_TARGET_EXT));
+#endif
+
+		LOG_DBG("USB phy enabled");
```
It is already prefixed with "udc_dwc2", maybe just `LOG_DBG("PHY enabled");`.

jfischer-no commented on samples/subsys/usb/mass/boards/esp32s3_devkitm_procpu.overlay:
```diff
@@ -0,0 +1,18 @@
+/*
+ * Copyright (c) 2025 Espressif Systems (Shanghai) Co., Ltd.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+/ {
+	msc_disk0 {
+		compatible = "zephyr,flash-disk";
+		partition = <&storage_partition>;
+		disk-name = "NAND";
+		cache-size = <4096>;
+	};
+};
+
+&zephyr_udc0 {
+	status = "okay";
+};
```
This kind of trivial overlays for all the Espressif boards is not acceptable. You must find another way to work around your hardware limitations. Maybe snippets can help you.

sylvioalves replied:
@jfischer-no By trivial, you mean only this snippet, right?
```c
&zephyr_udc0 {
	status = "okay";
};
```

raffarost replied:
As we commented previously, we can't leave USB OTG enabled in the board configuration. This would effectively disable USB-JTAG debugging out-of-the-box for any build, even for those not using USB-OTG. As USB-JTAG for debugging has to naturally be available without configurations to disable USB-OTG, we believe this is the right configuration for the board.

marekmatej replied:
> This kind of trivial overlays for all the Espressif boards is not acceptable. You must find another way to work around your hardware limitations. Maybe snippets can help you.

Yes I agree. Since having those on board level is limiting usage, I too think the `snippets` is probably the best place to have it.

raffarost commented:
@jfischer-no are there any points we can work on to move on with the PR?

jfischer-no commented on boards/espressif/esp32s3_devkitm/esp32s3_devkitm_procpu.dts:
```diff
@@ -151,3 +148,7 @@
 };
+
+zephyr_udc0: &usb_otg {
+	status = "disabled";
```
>Label zephyr_udc0 is necessary to run samples but it must remain
disabled as USB/JTAG shares the same phy and pins and won't work
for builds that don't use USB.

I am confused. The two boards have a dedicated USB connector for the USB controller (FS PHY). 

https://docs.zephyrproject.org/latest/boards/espressif/esp32s3_devkitc/doc/index.html
https://docs.zephyrproject.org/latest/boards/espressif/esp32s3_devkitm/doc/index.html

I accidentally found something that looks like an esp32s3_devkitm board. When I enable the default USB device controller in `boards/espressif/esp32s3_devkitm/esp32s3_devkitm_procpu.dts` and build a USB sample, I can flash and use the USB device simultaneously.

So why should it remain disabled?

raffarost replied:
You are right. I missed the fact that the internal transceiver would be switched off USB-JTAG only in case an enable operation is made, so it should be fine to keep them enabled by default. I updated the files.

tmon-nordic commented on snippets/espressif-usb-mass/snippet.yml:
```diff
@@ -0,0 +1,3 @@
+name: espressif-usb-mass
+append:
+  EXTRA_DTC_OVERLAY_FILE: espressif-usb-mass.overlay
```
`.overlay` in the sample folder is the status quo, so I think that's good enough for now. The real solution to the problem (samples requiring per board overlay) in my opinion should be out-of-scope of this PR.

hlord2000 commented:
> You need to enable the option CONFIG_USB_DEVICE_STACK_NEXT.

Thanks, it was enabled but USB_DEVICE_STACK was also enabled (verified in .config). Disabling USB_DEVICE_STACK fixed the compile error.

jfischer-no commented on boards/espressif/esp32s3_devkitc/esp32s3_devkitc_procpu.dts:
```diff
@@ -1,5 +1,5 @@
 /*
- * Copyright (c) 2024 Espressif Systems (Shanghai) Co., Ltd.
+ * Copyright (c) 2024-2025 Espressif Systems (Shanghai) Co., Ltd.
```
FYI, just `Copyright Espressif Systems (Shanghai) Co., Ltd.` is fine, see https://docs.zephyrproject.org/latest/contribute/guidelines.html#copyrights-notices

raffarost replied:
Since all our files follow this pattern I'll leave it as is for now.

jfischer-no commented on boards/espressif/esp32s3_devkitc/esp32s3_devkitc_procpu.dts:
```diff
@@ -29,11 +31,6 @@
 		zephyr,bt-hci = &esp32_bt_hci;
 	};
 
-	aliases {
-		uart-0 = &uart0;
-		sw0 = &button0;
-	};
-
```
These changes have nothing to do with "Add default zephyr_udc0 label for USB-OTG.". You may consider them harmless, but that is not how the commit should be. Please rework it and create a separate commit for the aliases property changes.

jfischer-no commented on samples/subsys/usb/mass/boards/esp32s3_devkitm_procpu.overlay:
This overlay is unnecessary, and I am pretty sure I already mentioned that. There are enough examples of how to use a board storage partition with a flashdisk driver.
